drivers/fpga/fpga_zynqmp.c

```diff
@@ -0,0 +1,323 @@
+ * The bitstream header has 5 sections starting with the letters a,b,c...
+ * Each section has the following structure:
+ * [key][length of data][data]
+ */
+static uint32_t *parse_header(const struct device *dev, uint32_t *image_ptr,
+		       uint32_t *img_size)
+{
+	unsigned char *header = (unsigned char *)image_ptr;
+
+	uint32_t length = LENGTH(header);
+
+	/* shift to the next section*/
+	header += 0x4U + length;
+
+	if (*header++ != 'a') {
+		LOG_ERR("Incorrect bitstream format");
+		return NULL;
+	}
+
+	length = LENGTH(header);
+	/* shift to the data section*/
+	header += 0x2U;
+
+	LOG_DBG("Design name = %s", header);
+
+	header += length;
+
+	if (*header++ != 'b') {
+		LOG_ERR("Incorrect bitstream format");
+		return NULL;
+	}
+
+	length = LENGTH(header);
+	/* shift to the data section*/
+	header += 0x2U;
+	LOG_DBG("Part name = %s", header);
+
+	header += length;
+
+	if (*header++ != 'c') {
+		LOG_ERR("Incorrect bitstream format");
+		return NULL;
+	}
+
+	length = LENGTH(header);
+	/* shift to the data section*/
+	header += 0x2U;
+
+	LOG_DBG("Date =  %s", header);
+
+	header += length;
+
+	if (*header++ != 'd') {
+		LOG_ERR("Incorrect bitstream format");
+		return NULL;
+	}
+
+	length = LENGTH(header);
+	/* shift to the data section*/
+	header += 0x2U;
+
+	LOG_DBG("Time =  %s", header);
+
+	header += length;
+
+	if (*header++ != 'e') {
+		LOG_ERR("Incorrect bitstream format");
+		return NULL;
+	}
+
+	/*
+	 * The last section is the raw bitstream.
+	 * It is preceded by its size, which is needed for DMA transfer.
+	 */
+	*img_size =
+		((uint32_t)*header << 24) | ((uint32_t) *(header + 1) << 16) |
+		((uint32_t) *(header + 2) << 8) | ((uint32_t) *(header + 3));
+
+	return (uint32_t *)header;
+}
+
+static int csudma_transfer(uint32_t size)
+{
+	/* setup the source DMA channel */
+	CSUDMA_SRC_ADDR = (uint32_t)BITSTREAM & CSUDMA_SRC_ADDR_MASK;
+	CSUDMA_SRC_ADDR_MSB = 0;
+	CSUDMA_SRC_SIZE = size << CSUDMA_SRC_SIZE_SHIFT;
+
+	/* wait for the SRC_DMA to complete */
+	while ((CSUDMA_SRC_I_STS & CSUDMA_I_STS_DONE_MASK) != CSUDMA_I_STS_DONE_MASK) {
+	};
+
+	/* acknowledge the transfer has completed */
+	CSUDMA_SRC_I_STS = CSUDMA_I_STS_DONE_MASK;
+
+	return 0;
+}
+
+static int wait_for_done(void)
+{
+	/* wait for PCAP PL_DONE */
+	while ((PCAP_STATUS & PCAP_PL_DONE_MASK) != PCAP_PL_DONE_MASK) {
+	};
+
+	PCAP_RESET = PCAP_RESET_MASK;
+	power_up_fpga();
+
+	return 0;
+}
+
+static enum FPGA_status zynqmp_fpga_get_status(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	if ((PCAP_STATUS & PCAP_PL_INIT_MASK) && (PCAP_STATUS & PCAP_PL_DONE_MASK)) {
+		return FPGA_STATUS_ACTIVE;
+	} else {
+		return FPGA_STATUS_INACTIVE;
+	}
+}
+
+static const char *zynqmp_fpga_get_info(const struct device *dev)
+{
+	struct zynqmp_fpga_data *data = dev->data;
+
+	return data->FPGA_info;
+}
+
+static int zynqmp_fpga_reset(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* Reset PL */
+	PCAP_PROG = PCAP_PROG_RESET_MASK;
+	PCAP_PROG = ~PCAP_PROG_RESET_MASK;
+
```

Comment:
shouldn't we check here if the PL is out of reset? `pcap_status` register provides such info.

Reply:
Yes, we should. I added a check to see if PL is out of reset. Thank you!

Reply:
thanks